src/layers/decoder_layer.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================================
+#include "decoder_layer.h"
+#include "attention.h"
+#include "kvcache_manager.h"
+#include "layers_attention.h"
+#include "layers_mlp.h"
+#include "mlp_llama.h"
+#include "rms_norm.h"
+
+#include <unordered_map>
+
+namespace xft {
+
+void invokeLayerLLaMA(DataType dt, ActivationType at, int batchSize, int inputSeqLen, int attHeadDim, int attHeadNum, int kvHeadNum,
+        int maxPositions, int maxPosEmbed, int pastSeqLen, int currentSeqLen, int step, int hiddenSize,
+        int intermediateSize, void *output, int outputStride, const void *input, int inputStride, const float *ln1Gamma,
+        const float *ln1Beta, const void *queryWeight, const void *keyWeight, const void *valueWeight,
+        const void *attnOutWeight, const float *ln2Gamma, const float *ln2Beta, const void *gateWeight,
+        const void *upWeight, const void *downWeight, const float *queryBias = nullptr, const float *keyBias = nullptr,
+        const float *valueBias = nullptr, const float *attnOutBias = nullptr) {
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lock(mutex);
+
+    auto prepareAttnMask = [&](DecoderContext *ctx, int step) {
+        int seqLen = ctx->inputSeqLen;
+        int accSeqLen = pastSeqLen + currentSeqLen;
+        float *mask = nullptr;
+
+        auto getAttnMask = [](int sizeRequired) {
```

Comment:
we are deprecating attention mask, so need to change this in future.

Reply:
Add TODO.